include/seqan3/search/kmer_index/shape.hpp

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \author Enrico Seiler <enrico.seiler AT fu-berlin.de>
+ * \author Vinzenz May <vinzenz.may AT fu-berlin.de>
+ * \brief Provides seqan3::shape.
+ */
+
+#pragma once
+
+#include <seqan3/range/container/small_vector.hpp>
+#include <seqan3/std/concepts>
+
+namespace seqan3
+{
+
+/*!\brief A class that defines what positions of a pattern to hash.
+*
+* When hashing a sequence, there may be positions that do not count towards the final hash value.
+* A shape offers an easy way to define such patterns. Given a k-mer length `k` (0 < `k` <= 32), a shape
+* represents a binary sequence where a `0` encodes a "don't care position", i.e. a position that is not taken into
+* account when computing the hash value. A `1` therefore translates to a position that is used to compute the hash
+* value.
+*
+* ### Example
+*
+* \include test/snippet/search/kmer_index/shape.cpp
+*
+* \attention 0 < size <= 32
+*/
+
+class shape : public small_vector<bool, 32>
+{
+public:
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr shape()                          noexcept = default; //!< Defaulted.
+    constexpr shape(shape const &)             noexcept = default; //!< Defaulted.
+    constexpr shape(shape &&)                  noexcept = default; //!< Defaulted.
+    constexpr shape & operator=(shape const &) noexcept = default; //!< Defaulted.
+    constexpr shape & operator=(shape &&)      noexcept = default; //!< Defaulted.
+    ~shape()                                   noexcept = default; //!< Defaulted.
+
+    /*!\brief Construct an ungapped shape from a given size.
+     *
+     * ### Complexity
+     *
+     * Linear in k.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     *
+     * \attention The size must be in the interval [1, 32]. In Debug mode, an assertion will check this constraint.
+     */
+    constexpr shape(size_t k) noexcept : small_vector<bool, 32>(k, true) {}
+
+    /*!\brief Construct from a parameter pack.
+     *
+     * ### Complexity
+     *
+     * Linear in the number of arguments.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     *
+     * \attention The size of the pack must be in the interval [1, 32].
+     */
+    template <typename... t>
+    //!\cond
+        requires sizeof...(t) >= 2 && sizeof...(t) <= 32 && (std::Integral<t> && ...)
+    //!\endcond
+    constexpr shape(t... args) noexcept
+    {
+        assert([] (auto... a) constexpr { return ((a == 0 || a == 1) && ...); }(args...)); // Only 0 and 1 allowed
```

Comment:
Can't you do 
```cpp
assert( ((args == 0) || (args == 1) && ...) );
```

Reply:
```
shape.hpp:87:30: error: binary expression in operand of fold-expression
   87 |         assert( ((args == 0) || (args == 1) && ...) );
```

🤷‍♂

Reply:
This will work:
```cpp
assert( ((args == 0 || args == 1) && ...) );
```
(the same syntax you use inside the lambda :wink: )